drivers/usb/usbuhci_new/usbuhci.c

```diff
@@ -0,0 +1,2791 @@
+                      IN PVOID uhciEndpoint,
+                      IN ULONG EndpointStatus)
+{
+    PUHCI_ENDPOINT UhciEndpoint = uhciEndpoint;
+    ULONG PhysicalAddress;
+
+    DPRINT("UhciSetEndpointStatus: uhciEndpoint - %p, EndpointStatus - %X\n",
+           uhciEndpoint,
+           EndpointStatus);
+
+    if (EndpointStatus != USBPORT_ENDPOINT_RUN)
+        return;
+
+    if (!(UhciEndpoint->Flags & UHCI_ENDPOINT_FLAG_HALTED))
+        return;
+
+    UhciEndpoint->Flags &= ~UHCI_ENDPOINT_FLAG_HALTED;
+
+    if (UhciEndpoint->HeadTD == NULL)
+        UhciEndpoint->TailTD = NULL;
+
+    if (UhciEndpoint->HeadTD)
+    {
+        PhysicalAddress = UhciEndpoint->HeadTD->PhysicalAddress;
+        PhysicalAddress &= ~UHCI_TD_LINK_PTR_TERMINATE;
+        UhciEndpoint->QH->HwQH.NextElement = PhysicalAddress;
+        UhciEndpoint->QH->HwQH.NextElement &= ~UHCI_QH_ELEMENT_LINK_PTR_QH;
+    }
+    else
+    {
+        UhciEndpoint->QH->HwQH.NextElement = UHCI_QH_ELEMENT_LINK_PTR_TERMINATE;
+    }
+}
+
+VOID
+NTAPI
+UhciPollIsoEndpoint(IN PUHCI_EXTENSION UhciExtension,
+                    IN PUHCI_ENDPOINT UhciEndpoint)
+{
+    DPRINT_IMPL("UhciPollIsoEndpoint: UNIMPLEMENTED. FIXME\n");
+}
+
+VOID
+NTAPI
+UhciPollNonIsoEndpoint(IN PUHCI_EXTENSION UhciExtension,
+                       IN PUHCI_ENDPOINT UhciEndpoint)
+{
+    PUHCI_HCD_QH QH;
+    PUHCI_HCD_TD NextTD;
+    PUHCI_HCD_TD TD;
+    ULONG NextTdPA;
+    ULONG PhysicalAddress;
+    SIZE_T TransferedLen;
+    PLIST_ENTRY ListTDs;
+    UCHAR TdStatus;
+
+    DPRINT_UHCI("UhciPollNonIsoEndpoint: UhciExtension - %p, UhciEndpoint - %p\n",
+                UhciExtension,
+                UhciEndpoint);
+
+    if (UhciEndpoint->Flags & UHCI_ENDPOINT_FLAG_HALTED)
+    {
+        DPRINT("UhciPollNonIsoEndpoint: Ep->Flags & UHCI_ENDPOINT_FLAG_HALTED \n");
+        return;
+    }
+
+    QH = UhciEndpoint->QH;
+
+    NextTdPA = QH->HwQH.NextElement & UHCI_QH_ELEMENT_LINK_POINTER_MASK;
+
+    if (NextTdPA)
+    {
+        NextTD = RegPacket.UsbPortGetMappedVirtualAddress(NextTdPA,
+                                                          UhciExtension,
+                                                          UhciEndpoint);
+    }
+    else
+    {
+        NextTD = NULL;
+    }
+
+    DPRINT_UHCI("UhciPollNonIsoEndpoint: NextTD - %p, NextTdPA - %p\n",
+                NextTD,
+                NextTdPA);
+
+    for ( TD = UhciEndpoint->HeadTD; ; TD = TD->NextHcdTD )
+    {
+        DPRINT_UHCI("UhciPollNonIsoEndpoint: TD - %p, TD->NextHcdTD - %p\n",
+                    TD,
+                    TD->NextHcdTD);
+
+        if (TD != NextTD && TD != NULL)
+        {
+            TD->Flags |= UHCI_HCD_TD_FLAG_DONE;
+            InsertTailList(&UhciEndpoint->ListTDs, &TD->TdLink);
+
+            if (TD->NextHcdTD &&
+                TD->NextHcdTD->HwTD.ControlStatus.Status & UHCI_TD_STS_ACTIVE)
+            {
+                if (NextTdPA == 0)   
+                    break;
```

Comment:
This is the only place that will run the code after the for loop. So that code should move here, then all the `goto ProcessListTDs;`s can become `break;`s